File: validate/semantic.go
package validate

import (
	"fmt"
	"path"
	"strings"

	"pocketknife/schema"
)

// semantic runs the checks that JSON Schema cannot express. It walks the parsed
// model and accumulates structured errors; it never stops at the first problem,
// so a caller sees every issue at once.
func semantic(app *schema.App) Errors {
	var errs Errors

	entityIDs := map[string]bool{}
	entityNames := map[string]bool{}

	for ei, ent := range app.Entities {
		epath := fmt.Sprintf("/entities/%d", ei)

		if entityIDs[ent.ID] {
			errs = append(errs, Error{epath + "/id", "duplicate_id", fmt.Sprintf("entity id %q is not unique", ent.ID)})
		}
		entityIDs[ent.ID] = true

		if entityNames[ent.Name] {
			errs = append(errs, Error{epath + "/name", "duplicate_name", fmt.Sprintf("entity name %q is not unique", ent.Name)})
		}
		entityNames[ent.Name] = true

		// An entity name becomes a URL segment under /apps/{app}/, so it must
		// not shadow a platform route. The id is checked too: a later rename
		// onto the id would otherwise smuggle the collision in.
		if isReservedEntity(ent.Name) {
			errs = append(errs, Error{epath + "/name", "reserved_name", fmt.Sprintf("entity name %q is reserved by the platform", ent.Name)})
		}
		if isReservedEntity(ent.ID) {
			errs = append(errs, Error{epath + "/id", "reserved_id", fmt.Sprintf("entity id %q is reserved by the platform", ent.ID)})
		}

		fieldIDs := map[string]bool{}
		fieldNames := map[string]bool{}

		for fi, f := range ent.Fields {
			fpath := fmt.Sprintf("%s/fields/%d", epath, fi)

			if fieldIDs[f.ID] {
				errs = append(errs, Error{fpath + "/id", "duplicate_id", fmt.Sprintf("field id %q is not unique within entity %q", f.ID, ent.Name)})
			}
			fieldIDs[f.ID] = true

			// A field id becomes a physical column name (D1), so it must not
			// collide with a platform-managed column.
			if isReserved(f.ID) {
				errs = append(errs, Error{fpath + "/id", "reserved_id", fmt.Sprintf("field id %q is reserved by the platform", f.ID)})
			}

			if fieldNames[f.Name] {
				errs = append(errs, Error{fpath + "/name", "duplicate_name", fmt.Sprintf("field name %q is not unique within entity %q", f.Name, ent.Name)})
			}
			fieldNames[f.Name] = true

			if isReserved(f.Name) {
				errs = append(errs, Error{fpath + "/name", "reserved_name", fmt.Sprintf("field name %q is reserved by the platform", f.Name)})
			}

			errs = append(errs, validateField(fpath, app, ent, f)...)
		}
	}

	fnIDs := map[string]bool{}
	fnNames := map[string]bool{}
	for fi, fn := range app.Functions {
		fpath := fmt.Sprintf("/functions/%d", fi)

		if fnIDs[fn.ID] {
			errs = append(errs, Error{fpath + "/id", "duplicate_id", fmt.Sprintf("function id %q is not unique", fn.ID)})
		}
		fnIDs[fn.ID] = true

		if fnNames[fn.Name] {
			errs = append(errs, Error{fpath + "/name", "duplicate_name", fmt.Sprintf("function name %q is not unique", fn.Name)})
		}
		fnNames[fn.Name] = true

		if fn.IsPrompt() {
			_, malformed := schema.ScanPrompt(fn.Prompt)
			for _, off := range malformed {
				errs = append(errs, Error{fpath + "/prompt", "malformed_placeholder", fmt.Sprintf("prompt of function %q contains a '{{' at byte %d that does not form a valid {{placeholder}} (there is no escaping)", fn.Name, off)})
			}
		} else {
			// The entry is joined onto the app directory and read at invoke
			// time, so it must stay inside the app directory.
			if path.IsAbs(fn.Entry) || fn.Entry != path.Clean(fn.Entry) || fn.Entry == ".." || strings.HasPrefix(fn.Entry, "../") {
				errs = append(errs, Error{fpath + "/entry", "bad_entry", fmt.Sprintf("function %q entry %q must be a clean relative path inside the app directory", fn.Name, fn.Entry)})
			}
		}

		errs = append(errs, validateCapabilities(fpath+"/capabilities", app, fn)...)
	}
	return errs
}

// validateCapabilities checks that every declared data scope resolves to a
// real entity in this manifest, is not repeated, and never requests an
// operation the entity itself does not allow — a scope the sandbox could
// never actually exercise is a manifest authoring mistake, not a runtime
// concern.
func validateCapabilities(path string, app *schema.App, fn *schema.Function) Errors {
	var errs Errors
	seenEntities := map[string]bool{}

	for di, ds := range fn.Capabilities.Data {
		dpath := fmt.Sprintf("%s/data/%d", path, di)

		if seenEntities[ds.Entity] {
			errs = append(errs, Error{dpath + "/entity", "duplicate_data_scope", fmt.Sprintf("entity %q is scoped more than once in function %q", ds.Entity, fn.Name)})
		}
		seenEntities[ds.Entity] = true

		ent := app.EntityByID(ds.Entity)
		if ent == nil {
			errs = append(errs, Error{dpath + "/entity", "unresolved_reference", fmt.Sprintf("data scope entity %q does not resolve to an entity in this manifest", ds.Entity)})
			continue
		}
		for _, op := range ds.Operations {
			if !ent.Allows(op) {
				errs = append(errs, Error{dpath + "/operations", "scope_exceeds_entity", fmt.Sprintf("function %q requests %q on entity %q, which the entity itself does not allow", fn.Name, op, ent.Name)})
			}
		}
	}

	seenDomains := map[string]bool{}
	for ni, d := range fn.Capabilities.Network {
		if seenDomains[d] {
			errs = append(errs, Error{fmt.Sprintf("%s/network/%d", path, ni), "duplicate_domain", fmt.Sprintf("domain %q is repeated", d)})
		}
		seenDomains[d] = true
	}
	return errs
}

func isReserved(name string) bool {
	for _, r := range schema.ReservedNames {
		if name == r {
			return true
		}
	}
	return false
}

func isReservedEntity(name string) bool {
	for _, r := range schema.ReservedEntityNames {
		if name == r {
			return true
		}
	}
	return false
}

// validateField checks min/max ordering, reference resolution, enum value
// integrity, and that any default satisfies the field's own constraints.
func validateField(path string, app *schema.App, ent *schema.Entity, f *schema.Field) Errors {
	var errs Errors

	if f.Min != nil && f.Max != nil && *f.Min > *f.Max {
		errs = append(errs, Error{path, "bad_bounds", fmt.Sprintf("min (%v) is greater than max (%v)", *f.Min, *f.Max)})
	}

	switch f.Type {
	case schema.TypeReference:
		if app.EntityByID(f.Target) == nil {
			errs = append(errs, Error{path + "/target", "unresolved_reference", fmt.Sprintf("reference target %q does not resolve to an entity in this manifest", f.Target)})
		}
	case schema.TypeEnum:
		seen := map[string]bool{}
		for _, v := range f.Values {
			if seen[v] {
				errs = append(errs, Error{path + "/values", "duplicate_enum_value", fmt.Sprintf("enum value %q is repeated", v)})
			}
			seen[v] = true
		}
	}

	if f.HasDefault {
		errs = append(errs, validateDefault(path+"/default", f)...)
	}
	return errs
}

// validateDefault confirms a declared default would itself pass field
// validation: enum membership, length bounds (text) and value bounds
// (integer/real).
func validateDefault(path string, f *schema.Field) Errors {
	var errs Errors
	switch f.Type {
	case schema.TypeText:
		s, ok := f.Default.(string)
		if !ok {
			return Errors{{path, "bad_default", "default must be a string"}}
		}
		n := float64(len([]rune(s)))
		if f.Min != nil && n < *f.Min {
			errs = append(errs, Error{path, "bad_default", fmt.Sprintf("default length %d is below min %v", len([]rune(s)), *f.Min)})
		}
		if f.Max != nil && n > *f.Max {
			errs = append(errs, Error{path, "bad_default", fmt.Sprintf("default length %d exceeds max %v", len([]rune(s)), *f.Max)})
		}
	case schema.TypeInteger:
		n, ok := f.Default.(int64)
		if !ok {
			return Errors{{path, "bad_default", "default must be an integer"}}
		}
		if f.Min != nil && float64(n) < *f.Min {
			errs = append(errs, Error{path, "bad_default", fmt.Sprintf("default %d is below min %v", n, *f.Min)})
		}
		if f.Max != nil && float64(n) > *f.Max {
			errs = append(errs, Error{path, "bad_default", fmt.Sprintf("default %d exceeds max %v", n, *f.Max)})
		}
	case schema.TypeReal:
		n, ok := f.Default.(float64)
		if !ok {
			return Errors{{path, "bad_default", "default must be a number"}}
		}
		if f.Min != nil && n < *f.Min {
			errs = append(errs, Error{path, "bad_default", fmt.Sprintf("default %v is below min %v", n, *f.Min)})
		}
		if f.Max != nil && n > *f.Max {
			errs = append(errs, Error{path, "bad_default", fmt.Sprintf("default %v exceeds max %v", n, *f.Max)})
		}
	case schema.TypeEnum:
		s, ok := f.Default.(string)
		if !ok {
			return Errors{{path, "bad_default", "default must be a string"}}
		}
		found := false
		for _, v := range f.Values {
			if v == s {
				found = true
				break
			}
		}
		if !found {
			errs = append(errs, Error{path, "bad_default", fmt.Sprintf("enum default %q is not one of the declared values", s)})
		}
	}
	return errs
}

File: agent/src/seams/semantic.ts
// Semantic checks JSON Schema structural validation cannot express, mirroring
// validate/semantic.go's rules closely enough that a manifest passing here
// will also pass the real gate: stable-id uniqueness, sibling-name
// uniqueness, reserved-name avoidance, reference resolution, enum value
// integrity, and defaults that satisfy the field's own constraints.

import type { AppManifest, Entity, Field } from "./manifest-types.js";
import { RESERVED_NAMES, entityById, fieldHasDefault } from "./manifest-types.js";
import type { ValidationError } from "./validator.js";

export function semanticErrors(manifest: AppManifest): ValidationError[] {
  const errors: ValidationError[] = [];

  const entityIds = new Set<string>();
  const entityNames = new Set<string>();

  manifest.entities.forEach((entity, ei) => {
    const epath = `/entities/${ei}`;

    if (entityIds.has(entity.id)) {
      errors.push({ path: `${epath}/id`, message: `entity id "${entity.id}" is not unique` });
    }
    entityIds.add(entity.id);

    if (entityNames.has(entity.name)) {
      errors.push({ path: `${epath}/name`, message: `entity name "${entity.name}" is not unique` });
    }
    entityNames.add(entity.name);

    const fieldIds = new Set<string>();
    const fieldNames = new Set<string>();

    entity.fields.forEach((field, fi) => {
      const fpath = `${epath}/fields/${fi}`;

      if (fieldIds.has(field.id)) {
        errors.push({
          path: `${fpath}/id`,
          message: `field id "${field.id}" is not unique within entity "${entity.name}"`,
        });
      }
      fieldIds.add(field.id);

      if (RESERVED_NAMES.includes(field.id)) {
        errors.push({ path: `${fpath}/id`, message: `field id "${field.id}" is reserved by the platform` });
      }

      if (fieldNames.has(field.name)) {
        errors.push({
          path: `${fpath}/name`,
          message: `field name "${field.name}" is not unique within entity "${entity.name}"`,
        });
      }
      fieldNames.add(field.name);

      if (RESERVED_NAMES.includes(field.name)) {
        errors.push({ path: `${fpath}/name`, message: `field name "${field.name}" is reserved by the platform` });
      }

      errors.push(...validateField(fpath, manifest, entity, field));
    });
  });

  const fnIds = new Set<string>();
  const fnNames = new Set<string>();
  (manifest.functions ?? []).forEach((fn, fi) => {
    const fpath = `/functions/${fi}`;

    if (fnIds.has(fn.id)) {
      errors.push({ path: `${fpath}/id`, message: `function id "${fn.id}" is not unique` });
    }
    fnIds.add(fn.id);

    if (fnNames.has(fn.name)) {
      errors.push({ path: `${fpath}/name`, message: `function name "${fn.name}" is not unique` });
    }
    fnNames.add(fn.name);

    (fn.capabilities.data ?? []).forEach((scope, di) => {
      if (entityById(manifest, scope.entity) === undefined) {
        errors.push({
          path: `${fpath}/capabilities/data/${di}/entity`,
          message: `data scope entity "${scope.entity}" does not resolve to an entity in this manifest`,
        });
      }
    });
  });

  return errors;
}

function validateField(path: string, manifest: AppManifest, _entity: Entity, field: Field): ValidationError[] {
  const errors: ValidationError[] = [];

  if ("min" in field && "max" in field && field.min !== undefined && field.max !== undefined && field.min > field.max) {
    errors.push({ path, message: `min (${field.min}) is greater than max (${field.max})` });
  }

  if (field.type === "reference") {
    if (entityById(manifest, field.target) === undefined) {
      errors.push({
        path: `${path}/target`,
        message: `reference target "${field.target}" does not resolve to an entity in this manifest`,
      });
    }
  }

  if (field.type === "enum") {
    const seen = new Set<string>();
    field.values.forEach((v) => {
      if (seen.has(v)) {
        errors.push({ path: `${path}/values`, message: `enum value "${v}" is repeated` });
      }
      seen.add(v);
    });
  }

  if (fieldHasDefault(field)) {
    errors.push(...validateDefault(`${path}/default`, field));
  }

  return errors;
}

function validateDefault(path: string, field: Field): ValidationError[] {
  switch (field.type) {
    case "text": {
      const n = field.default!.length;
      const errors: ValidationError[] = [];
      if (field.min !== undefined && n < field.min) {
        errors.push({ path, message: `default length ${n} is below min ${field.min}` });
      }
      if (field.max !== undefined && n > field.max) {
        errors.push({ path, message: `default length ${n} exceeds max ${field.max}` });
      }
      return errors;
    }
    case "integer":
    case "real": {
      const n = field.default!;
      const errors: ValidationError[] = [];
      if (field.min !== undefined && n < field.min) {
        errors.push({ path, message: `default ${n} is below min ${field.min}` });
      }
      if (field.max !== undefined && n > field.max) {
        errors.push({ path, message: `default ${n} exceeds max ${field.max}` });
      }
      return errors;
    }
    case "enum": {
      if (!field.values.includes(field.default!)) {
        return [{ path, message: `enum default "${field.default}" is not one of the declared values` }];
      }
      return [];
    }
    default:
      return [];
  }
}
